registry/cache/warming.go

```diff
@@ -273,6 +273,10 @@ func (w *Warmer) warm(ctx context.Context, now time.Time, logger log.Logger, id
 		awaitFetchers := &sync.WaitGroup{}
 		awaitFetchers.Add(len(toUpdate))
 
+		ctxc, cancel := context.WithCancel(context.Background())
```

Comment:
This should be created with the context passed into the func; otherwise, any deadline or cancellation signal given to that context will be ignored.

Reply:
You should also use this fresh context for the requests, i.e., in L304 below. Otherwise you have a race between the context being cancelled, and the early exit.

Reply:
And for the loop-breaking select in L282-286.